### PR TITLE
metrics: Show top CPU core usage

### PR DESCRIPTION
**Pull request by martinpitt**
Often some application maxes out and gets bottle-necked on a single CPU
core. On systems with a lot of CPUs this is not visible in the average
CPU progress though. It is visible in the tooltip, but first this is not
very discoverable, and second there is no indication for the user to
actually check this, as there might be dozens of idle cores next to the
single 100% one.

On multi-core systems, show a second CPU progress bar for the currently
busiest CPU core. This provides a good balance between pointing out the
problem and avoiding cluttered UI with showing all of them all the time
(which is not even all that interesting).

Also add an info icon to the "n CPUs" progress if there are multiple
cores, so that users get a hint that there is a tooltip.

## Metrics: Show busiest CPU core

On multi-core systems, the Metrics page now shows an additional usage bar for the CPU core with the highest usage. This points out bottlenecks and performance problems which happen because only a few cores (or even a single one) get fully used, while many others are mostly idle. In that case, the average CPU usage will be low.

image

### CODE REVIEW

**Comment by martinpitt** [comment]
This also looks reasonable on mobile:

top-core-mobile

**Comment by marusak** [comment]
Nice! A few comments/questions for UI:
- Should we add (?) icon next to `8 CPUs` to make it obvious you can hover over it?
- Isn't this too jumpy? Meaning that the value of the core changes every second?
- I feel like people will not understand why we show random core there. I would label it something like `Busiest core (2)`. That would make it less jumpy as well as only tiny part would change.

**Review by garrett — CHANGES_REQUESTED** [review]
> Should we add (?) icon next to 8 CPUs to make it obvious you can hover over 

It's not hover help, however.

I looked through the PF icon list and don't see anything that's so obvious as an icon. 

There's  pf-icon-resources-empty, pf-icon-resources-almost-empty, pf-icon-resources-almost-full, pf-icon-resources-full, but that's kind of repeating the bar below it.

We could have a warning icon when load is high. It's sometimes unwanted. But the issue with this is that some people specifically _want_ a high load, so it's not a warning in their case.

Even with using an icon, it could be jumpy.

> Isn't this too jumpy? Meaning that the value of the core changes every second?

Exactly my concern.

> I feel like people will not understand why we show random core there. I would label it something like Busiest core (2). That would make it less jumpy as well as only tiny part would change.

Another big concern I have.

---

Also, what happens when more than one CPU core is being used a lot? Are they all shown?

And is this utilization (which can be good) or saturation? 

And then, what can someone actually do with this information? "Oh, core 2 is being used a lot... and now what?"

**Comment by garrett** [comment]
If we draw attention to the CPU to hover it, we could use something like fa-microchip ("Represents the CPU of a device") or some kind of tip icon (that doesn't exist in PF). A lightbulb is often used as a tip (but isn't part of the PF icons). We could have an ellipsis inside of a circle (...) to indicate additional details (but this icon also doesn't exist in PF).

I could make these icons and suggest them for upstream too, if any of them make sense here.

**Comment by martinpitt** [comment]
> Isn't this too jumpy? Meaning that the value of the core changes every second?

Yes, it can. See below.

> Also, what happens when more than one CPU core is being used a lot? Are they all shown?

No, that's exactly what I want to avoid. The main piece of feedback that we heard is that people on large servers (think, 128 CPUs or more, but it already happens with my 8 cores) don't actually see a "red" usage bar if only one CPU is 100%. But I absolutely don't want to draw 128 progress bars -- that breaks both the UI and also the user's attention, and is not even what they want to know.

- (Too) high average CPU usage (distributed over all cores) → it's time to shift some load to other servers
- average CPU load is fine, but my bitcoin mininig process is still slow → bottle-necked on a single CPU

Of course there can also be the case that there are five single-threaded processes which are all bottle-necked, but then we will still have a 100% top CPU core, and thus point out the problem that there *is* a bottleneck even though you have enough CPU power in total.

> And is this utilization (which can be good) or saturation?

It's utilization. Saturation is not defined in terms of "per-CPU core", as it's by definition the number of tasks which are waiting to be scheduled to a core.

> And then, what can someone actually do with this information? "Oh, core 2 is being used a lot... and now what?"

It's about the fact that the busiest CPU utilization diverges so much from the average. It's not really about "core 2" in particular -- I'm even happy to entirely drop this number, as it's by and large meaningless, and not connected to anything else in the UI. This would also make the label less "jumpy". I.e. it would then show:

> 8 CPUs --- 15%
> Busiest CPU: 100%

WDYT?

File: pkg/metrics/metrics.jsx
Comment (martinpitt) [inline pkg/metrics/metrics.jsx:463]: It should probably say "CPU", not "Core", as we call it CPU everywhere else. "Busiest CPU" sounded a bit strange to me. Bikeshedding appreciated!

**Comment by garrett** [comment]
I made a mockup thingy:
https://codepen.io/garrett/pen/MWEKovX

It looks like this:

image

You can mostly disregard the HTML/CSS — and the colors are just some named CSS ones; they're not even pretending to be accurate. The important part is that it groups the bars, stretches them to fit the container, and sets a min-height of 1px. This does make the progress bar stack grow when there are many CPUs, but it does grow reasonably.

Alternatively, we could do something that does focus on the average and calls out the max, such as:
https://codepen.io/garrett/pen/VwMeMbG

image
(We could make the link be a popover that shows all CPUs, in order, as opposed to a hover that shows an unordered list with percentages only.)

**Comment by martinpitt** [comment]
I like the second variant better, with just showing average and max (i.e. merging the two progress bars, essentially). Showing all the CPUs is way too jittery, and IMHO does not even contain useful information.

**Comment by garrett** [comment]
We do label the bars together as a group, but could also make sure there are tooltips when hovering over each. Perhaps "Combined average CPU usage: XX%" and "Maximum CPU usage: XX%"... and changing the view all CPUs to a popover could be a separate change, but probably makes sense, as we can visualize that a bit better than just a hover, and this is probably more discoverable.

Hopefully we can upgrade to PF >= 4.161.0 soon and add the `pf-m-tabular-nums` opt-in class so that the label wouldn't jump around much (except from X to XX to XXX although I guess we could have leading zeros for single or double digits and/or fix the width with CSS).

**Comment by martinpitt** [comment]
Independent issue: [TestCurrentMetrics.testMemory](https://logs.cockpit-project.org/logs/pull-16686-20211206-123627-ca37ceaf-fedora-35/log.html#291-1) fails on the second retry fairly reliably on the infrastructure. I can reproduce this locally after some attempts, will send a PR.

**Comment by martinpitt** [comment]
I locally made a version with a "merged" progress bar. On wide screens that looks quite decent:

image

However, on narrow windows it becomes too crammed, without any leftover space for an additional "View all CPUs":

cpu

and sometimes completely ridiculous:

image

We could (and should) of course adjust the minimum width of the card, but unless we make it really wide, it's still way too much text. It'll most probably get worse with translations.

IMHO it's a lot more relaxed (and also a lot less custom CSS/less ambiguity about two merged progress bars) with two separate ones:

image

image

**Comment by garrett** [comment]
This PR made me realize we have too much spacing on desktop modes and that we're sometimes lacking a space between "load" and the values, so I adjusted all that and sent a PR: https://github.com/cockpit-project/cockpit/pull/16692

**Comment by martinpitt** [comment]
@garrett : I updated the code a bit for the discussion above, and updated screenshots in the description and first comment. I'd appreciate a review here, in particular how we can point out the existence of the tooltip a bit more. Or of course we could also convert this to an actual [Popover](https://www.patternfly.org/v4/components/popover/)

**Comment by garrett** [comment]
Depending on the layout, gap might actually need to be x and y, not just 1 value, so it might need to be something like `gap: 0 var(--pf-global--spacer-sm)` so that it's 0 vertical and half spacer for horizontal.

**Comment by martinpitt** [comment]
The proposed narrow layout is really difficult, this isn't at all how the progress component is designed. So this will require some intrusive hacking into its CSS. I'm not sure if I'm up to that, but I tried to warm up with fixing the line breaks in "Load:" in PR #16867

**Comment by martinpitt** [comment]
@garrett : I updated this PR to a close version of your design. Moving the "average/max" *below* the "n CPUs" label is hard, that's just not how the Progress component is designed. But at least it flexes reasonably well now.

One thing which I still need to fix is to drop that icon if the CPU usage becomes > 90%:

image

That just looks awful, and I think we don't want an icon there. In principle that affects the single-CPU bar as well, so it'll be a separate commit. But with the new "max" view it's a bit more likely to encounter it, and it breaks the "gap-less stack" layout.

**Comment by garrett** [comment]
- I ported the stacked flexes to use a grid instead, so they're aligned relative to each other. 
- I sorted the CPUs, so they don't jump around.
- I made percentages use the fixed width font and aligned them to the right so it's easier to see which CPU has higher usage.
- I worked on the layout in general and had to use CSS to do it, as PF didn't give us what we needed.
- I also made the font smaller to match the other text in the cards.

Meanwhile, I noticed that sometimes the average CPU is larger than the maximum CPU usage:

image

Weird, right?

File: pkg/metrics/metrics.jsx
Comment (garrett) [inline pkg/metrics/metrics.jsx:466]: Oh, whoops. I think this is not supposed to be here... going to push a fix.

File: pkg/metrics/metrics.jsx
Comment (garrett) [inline pkg/metrics/metrics.jsx:463]: I hope I got this right. I guess the span isn't needed as I'm overriding it in the CSS anyway. I'll push a fix.

File: pkg/metrics/metrics.jsx
Comment (garrett) [reply pkg/metrics/metrics.jsx]: I could fix this commit, but it'll be squashed with the next one (and probably all the ones above) before this PR is merged.

File: pkg/metrics/metrics.jsx
Comment (martinpitt) [reply pkg/metrics/metrics.jsx:462]: The structure looks fine to me. However, the .sort()ing needs to go. `top_cores` already got sorted by descending usage two lines above, and it's been like that since day one of the popup. If we want to change that, I propose to do that in a separate PR (or at least commit).

File: pkg/metrics/metrics.jsx
Comment (martinpitt) [reply pkg/metrics/metrics.jsx]: Right, I squashed the commits.

**Comment by martinpitt** [comment]
I fixed the conflict, squashed the commits, and updated the main screenshot for the relnotes, and the popover screenshot.

**Review by garrett — CHANGES_REQUESTED** [review]
I think we're almost there.

Somehow, there's too much space between the CPU and load average:

image

And the load average is oddly spaced when wrapped:

image

I wonder if I caused these regressions in the flex adjustment stuff. (Did that land yet? I don't think so?) Or if something in this PR is affecting it.

Between the spacing issues and the sorting, I think we could either add fixes and have the sorting as a separate commit... but I'm happy if we merge this as-is and have follow-ups if you prefer (and I'll change my approval to approve if so).

**Review by garrett — APPROVED** [review]
Marking as approve.

---

I looked at the spacing issues, and they're my fault, I think... but only because we depended on PF's incorrect behavior when implementing the Flex widgets here. It should be super easy to fix though. But it should be done outside this PR.

FWIW: We are also importing duplicates of CSS:

Screenshot from 2022-03-01 11-37-04

We can either do the sorting change in a separate commit here or a follow-up PR.

**Comment by martinpitt** [comment]
@garrett Thanks for the review, and your hand-holding/improvements here! Can I leave the flex bits to you?

**Comment by garrett** [comment]
For reference, the "flex bits" were adjusted in 
- https://github.com/cockpit-project/cockpit/pull/17075

**Comment by garrett** [comment]
The screenshot needs to be updated for the release notes.

**Comment by martinpitt** [comment]
@garrett : Done, thanks for catching!

**Pull request merged**
